crates/typst/src/layout/grid/layout.rs

```diff
@@ -198,44 +317,156 @@ impl CellGrid {
+                    // pending hlines later would change their order!
+                    pending_hlines.push((span, line));
+                    continue;
+                }
+                GridItem::VLine { x, start, end, stroke, span, position } => {
+                    let x = x.as_custom().unwrap_or_else(|| {
+                        // When no 'x' is specified for the vline, we place it
+                        // after the latest automatically positioned cell.
+                        // The current value of the auto index is always the
+                        // index of the latest automatically positioned cell
+                        // placed plus one (that's what we do in
+                        // 'resolve_cell_position'), so we subtract 1 to get
+                        // that cell's index, and place the vline after its
+                        // column. The exception is when the auto_index is 0,
+                        // meaning no automatically positioned cell was placed
+                        // yet. In that case, we place the vline to the left of
+                        // the table.
+                        auto_index
+                            .checked_sub(1)
+                            .map_or(0, |last_auto_index| last_auto_index % c + 1)
+                    });
+                    if end.is_some_and(|end| end.get() < start) {
+                        bail!(span, "line cannot end before it starts");
+                    }
+                    let (x, line) = if position == LinePosition::After
+                        && (!has_gutter || x + 1 == c)
+                    {
+                        // Just place the line before the next column if
+                        // there's no gutter and the line should be placed
+                        // after the one with given index.
+                        // Note that placing after the last column is also the
+                        // same as just placing on the grid's end border, even
+                        // with gutter.
+                        (
+                            x + 1,
+                            Line {
+                                index: x + 1,
+                                start,
+                                end,
+                                stroke,
+                                position: LinePosition::Before,
+                            },
+                        )
+                    } else {
+                        (x, Line { index: x, start, end, stroke, position })
+                    };
+                    // There must exist at least one more column for this line
+                    // to be actually drawn if the "After" position was chosen.
+                    let necessary_cols =
+                        if line.position == LinePosition::After { x + 1 } else { x };
+                    if necessary_cols > c {
+                        bail!(span, "cannot place vertical line at invalid column");
```

Comment:
> It would be great if the invalid row/column error message would specify at which exact row/column it couldn't place things.

Agreed.

> Also, `table.vline(stroke: green, position: end)` at non-first row yields
> 
> > cannot place vertical line at invalid column.
> 
> Maybe it makes sense, but it struck me as a bit weird.

It's because you can only _automatically_ place at `x: 0` before the first cell. After the first row, a vline placed after the last cell of the row is, to the code, the same as a vline placed before the first cell of the next row, since, in the end, it's all just a vector of stuff. So I had to either choose one, or make it so placing consecutive vlines would step the `x` counter (such that the first vline would be at `x: c` and the second vline would be at `x: 0`) - the latter is tablex's current approach. I chose the first approach - last vline - because, well, otherwise you couldn't automatically specify a vline at the end (except, I guess, after the last row). Additionally, it felt like the extra magic wasn't worth it, and it just made tablex's inner algorithm more complicated, IMO. So I preferred it this way.

But the error could certainly be improved to mention that you were trying to place right after the border, for example.

Reply:
I made the error report the invalid position now (if the index is invalid), or that you're trying to place at the border but at the "after" position (if that's the case instead).